File: app/src/common/helpers/conceptMaps.ts
import type { IConceptMap } from "@ahryman40k/ts-fhir-types/lib/R4";

import type { Mapping } from "services/api/generated/api.generated";

export type CustomConceptMap = Record<string, string>;
export type CustomConceptMapById = Record<string, CustomConceptMap | undefined>;

export const getConceptMapIdsFromMappings = (mappings: Mapping): string[] =>
  mappings.resources?.reduce(
    (acc, resource) =>
      resource.attributes
        ? [
            ...acc,
            ...resource.attributes.reduce(
              (acc, attribute) =>
                attribute.input_groups
                  ? [
                      ...acc,
                      ...attribute.input_groups.reduce(
                        (acc, input_group) =>
                          input_group.inputs
                            ? [
                                ...acc,
                                ...input_group.inputs.reduce(
                                  (acc, input) =>
                                    input.concept_map_id
                                      ? [...acc, input.concept_map_id]
                                      : acc,
                                  [] as string[]
                                ),
                              ]
                            : acc,
                        [] as string[]
                      ),
                    ]
                  : acc,
              [] as string[]
            ),
          ]
        : acc,
    [] as string[]
  ) ?? [];

/**
 * getConceptMappingFromResource extracts the mapping from a ConceptMap resource
 * @param conceptMap
 * @returns a simple object with the concept mapping
 * Example: { "F": "female", "M": "male" }
 * we only handle a single target for each source
 */
export const getConceptMappingFromResource = (
  conceptMap: IConceptMap
): CustomConceptMap | undefined =>
  conceptMap.group?.reduce(
    (acc, group) => ({
      ...group.element.reduce(
        (acc, el) =>
          el.code && el.target?.[0]?.code
            ? { [el.code]: el.target?.[0]?.code, ...acc }
            : acc,
        {} as CustomConceptMap
      ),
      ...acc,
    }),
    {} as CustomConceptMap
  );

/**
 * mergeConceptMapsToMappings associates the transformed concept maps to their Inputs
 * @param mappings
 * @param conceptMapsById
 * @returns a mapping with conceptmaps
 */
export const mergeConceptMapsToMappings = (
  mappings: Mapping,
  conceptMapsById: CustomConceptMapById
): Mapping => ({
  ...mappings,
  resources: mappings.resources?.map((resource) => ({
    ...resource,
    attributes: resource.attributes?.map((attribute) => ({
      ...attribute,
      input_groups: attribute.input_groups?.map((input_group) => ({
        ...input_group,
        inputs: input_group.inputs?.map((input) =>
          input.concept_map_id
            ? {
                ...input,
                concept_map: conceptMapsById[input.concept_map_id],
              }
            : { ...input }
        ),
      })),
    })),
  })),
});

File: django/river/common/mapping/concept_maps.py
from typing import Optional

from common.adapters.fhir_api import fhir_api


# FIXME: use this function on a mapping before sending it to the analyzer
def dereference_concept_map(mapping, auth_token: Optional[str]):
    for attribute in mapping["attributes"]:
        for input_group in attribute["input_groups"]:
            for sql_input in input_group["sql_inputs"]:
                if concept_map_id := sql_input.get("concept_map_id"):
                    concept_map = format_concept_map(concept_map_id, auth_token)
                    sql_input["concept_map"] = concept_map


def format_concept_map(concept_map_id: str, auth_token: Optional[str]):
    concept_map_resource = fhir_api.retrieve("ConceptMap", concept_map_id, auth_token)
    concept_map = {}
    for group in concept_map_resource["group"]:
        for element in group["element"]:
            # NOTE we only handle a single target for each project
            concept_map[element["code"]] = element["target"][0]["code"]
    return concept_map
